src/Common/ProxyConfigurationResolverProvider.cpp

```diff
@@ -0,0 +1,189 @@
+
+        if (include_https_uris && configuration.has(config_prefix + ".https"))
+        {
+            auto https_uris = extractURIList(config_prefix + ".https", configuration);
+            uris.insert(uris.end(), https_uris.begin(), https_uris.end());
+        }
+
+        return uris.empty() ? nullptr : std::make_shared<ProxyListConfigurationResolver>(uris);
+    }
+
+    std::shared_ptr<ProxyConfigurationResolver> getListResolverOldSyntax(
+        const String & config_prefix,
+        const Poco::Util::AbstractConfiguration & configuration
+    )
+    {
+        auto uris = extractURIList(config_prefix, configuration);
+
+        return uris.empty() ? nullptr : std::make_shared<ProxyListConfigurationResolver>(uris);
+    }
+
+    std::shared_ptr<ProxyConfigurationResolver> getListResolver(
+        ProxyConfiguration::Protocol protocol, const String & config_prefix, const Poco::Util::AbstractConfiguration & configuration
+    )
+    {
+        std::vector<String> keys;
+        configuration.keys(config_prefix, keys);
+
+        bool new_setting_syntax = std::find_if(
+                                        keys.begin(),
+                                        keys.end(),
+                                        [](const String & key)
+                                        {
+                                            return startsWith(key, "http") || startsWith(key, "https");
+                                        }) != keys.end();
+
+        return new_setting_syntax ? getListResolverNewSyntax(protocol, config_prefix, configuration)
+                                  : getListResolverOldSyntax(config_prefix, configuration);
+    }
+}
+
+std::shared_ptr<ProxyConfigurationResolver> ProxyConfigurationResolverProvider::get(Protocol protocol)
+{
+    if (auto context = Context::getGlobalContextInstance())
+    {
+        if (auto resolver = getFromSettings(protocol, "", context->getConfigRef()))
+        {
+            return resolver;
+        }
+    }
+
+    return std::make_shared<EnvironmentProxyConfigurationResolver>(protocol);
+}
+
+std::shared_ptr<ProxyConfigurationResolver> ProxyConfigurationResolverProvider::getFromSettings(
+    Protocol protocol,
+    const String & config_prefix,
+    const Poco::Util::AbstractConfiguration & configuration
+)
+{
+    auto proxy_prefix = config_prefix.empty() ? "proxy" : config_prefix + ".proxy";
+
+    if (configuration.has(proxy_prefix))
+    {
+        std::vector<String> config_keys;
+        configuration.keys(proxy_prefix, config_keys);
+
+        if (auto resolver_configs = std::count_if(config_keys.begin(), config_keys.end(), [](const std::string & k) { return startsWith(k, "resolver"); }))
+        {
+            if (resolver_configs > 2)
+            {
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "Only two remote proxy resolvers are allowed, one for HTTP and one for HTTPs");
```

Comment:
Looks like it allows having two `http` (or two `https`) proxies configured, but doesn't allow, for example, 2 `http` + 1 `https`

Reply:
Hm.. indeed. Correct behavior should be: 1 HTTP and 1 HTTPs. Will fix that tomorrow

Reply:
@tavplubix what do you think about allowing the user to specify N HTTP remote resolvers and N HTTPs remote resolvers, but we only pick the first one of each and simply ignore the rest. It'll make code a lot simpler.

Reply:
Another thing that might be worth discussing is that these exceptions are not thrown at startup time, rather upon request.

In this case, the request will be aborted. Should we try the other proxy methods first (list and environment)?

Reply:
> allowing the user to specify N HTTP remote resolvers and N HTTPs remote resolvers, but we only pick the first one of each and simply ignore the rest.

It's okay

> Another thing that might be worth discussing is that these exceptions are not thrown at startup time, rather upon request.

It's okay that requests will fail if the server is misconfigured (it would be better to check on startup, but it's not necessary)

> Should we try the other proxy methods first (list and environment)?

No, we should not try to continue working if the server is misconfigured